src/classes/RD_RecurringDonations_TEST2.cls

```diff
@@ -1154,4 +1154,194 @@ private with sharing class RD_RecurringDonations_TEST2 {
+        r1.npe03__Contact__c = c.Id;
+        r1.npe03__Amount__c = 100;
+        r1.npe03__Date_Established__c = Date.newInstance(1970,6,12);
+        r1.npe03__Open_Ended_Status__c = System.label.npe03.RecurringDonationOpenStatus;
+        r1.npe03__Next_Payment_Date__c = System.today().toStartOfMonth();
+        r1.OwnerId = System.Userinfo.getUserId();
+        insert r1;        
+        
+        //--------------------------------
+        // Setup the Second Recurring Donation
+        npe03__Recurring_Donation__c r2 = new npe03__Recurring_Donation__c();
+        r2.Name = 'TestRD2';
+        r2.npe03__Installments__c = 0;
+        r2.npe03__Contact__c = c.Id;
+        r2.npe03__Amount__c = 100;
+        r2.npe03__Date_Established__c = Date.newInstance(1970,6,12);
+        r2.npe03__Open_Ended_Status__c = System.label.npe03.RecurringDonationOpenStatus;
+        r2.npe03__Next_Payment_Date__c = System.today().toStartOfMonth();
+        r2.OwnerId = System.Userinfo.getUserId();
+        insert r2;    
+
+        //--------------------------------
+        // Setup the Opportunity
+        String closedstage = [select masterlabel from opportunitystage where isActive = true and iswon = true and isClosed = true limit 1].masterlabel;
+        
+        Opportunity o = new Opportunity();
+        o.StageName = closedstage;
+        o.CloseDate = system.today();
+        o.Amount = 50;
+        o.AccountId = a.Id;
+        o.Name = 'test';
+        o.npe03__Recurring_Donation__c = r1.Id;
+        RD_ProcessControl.hasRun = false;
+        
+        //--------------------------------
+        // Insert the Opportunity
+        Test.startTest();        
+        insert o;
+        
+        //--------------------------------
+        // Confirm the Opportunity has updated only the Recurring Donation's (r1) npe03__Paid_Amount__c
+        npe03__Recurring_Donation__c rd1 = [select npe03__Paid_Amount__c from npe03__Recurring_Donation__c where id = :r1.id];    
+        System.assertEquals(50, rd1.npe03__Paid_Amount__c);
+
+        npe03__Recurring_Donation__c rd2 = [select npe03__Paid_Amount__c from npe03__Recurring_Donation__c where id = :r2.id];    
+        System.assertEquals(null, rd2.npe03__Paid_Amount__c);
+        
+        //--------------------------------
+        // Update the Second Opportunity's Recurring Donation field
+        o.npe03__Recurring_Donation__c = r2.Id;
+        update o;
+        Test.stopTest();
+
+        //--------------------------------
+        // Confirm the Opportunity has updated the npe03__Paid_Amount__c for both Recurring Donations (r1 & r2) 
+        rd1 = [select npe03__Paid_Amount__c from npe03__Recurring_Donation__c where id = :r1.id];    
+        System.assertEquals(null, rd1.npe03__Paid_Amount__c);
+
+        rd2 = [select npe03__Paid_Amount__c from npe03__Recurring_Donation__c where id = :r2.id];    
+        System.assertEquals(50, rd2.npe03__Paid_Amount__c);
+    }
+    
 }
```

Comment:
your first test covers an opp changing from no RD to an existing RD.  your second test covers moving an opp from RD1 to RD2.  but I think you missed testing an opp changing from an RD to no RD.  not sure if that would have then flagged the issues I mentioned above, or whether we are getting luck in the code that there is no negative side effect of having nulls in our set rdIds.  we should add the test and make the code avoid this.  thanks.